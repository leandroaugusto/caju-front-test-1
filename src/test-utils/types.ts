export interface ITestUtils {
  children?: React.ReactNode
}
