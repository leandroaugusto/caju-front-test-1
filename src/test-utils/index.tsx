import { ReactElement } from "react";
import { ThemeProvider } from "styled-components";
import { render, RenderOptions, RenderResult } from "@testing-library/react";

import { ITestUtils } from "./types";

import theme from "~/theme";

const AllTheProviders: React.FC = ({ children }: ITestUtils) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options,
  });

export * from "@testing-library/react";

export { customRender };
