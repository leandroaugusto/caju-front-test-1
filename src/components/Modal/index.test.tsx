import { vi } from "vitest";

import { customRender, fireEvent, screen } from "~/test-utils";
import { Modal } from ".";

describe("Modal", () => {
  const mockOnClose = vi.fn();
  const mockConfirm = vi.fn();

  const defaultProps = {
    open: true,
    onClose: mockOnClose,
    message: "Mock message",
    confirm: mockConfirm,
  };

  afterEach(() => {
    vi.clearAllMocks();
  });

  it("should render correctly when open", () => {
    const { container } = customRender(<Modal {...defaultProps} />);

    const modal = screen.getByTestId("modal-container");
    const message = screen.getByText("Mock message");
    const modalBackdrop = screen.getByTestId("modal-backdrop");
    const modalContent = screen.getByTestId("modal-content");

    expect(modal).toBeInTheDocument();
    expect(message).toBeInTheDocument();

    expect(modalBackdrop).toHaveStyle(
      "opacity: 1; visibility: visible; pointer-events: all;"
    );
    expect(modalContent).toHaveStyle(
      "opacity: 1; visibility: visible; pointer-events: all;"
    );

    expect(container).toMatchSnapshot();
  });

  it("should does not render when not open", () => {
    const { queryByTestId } = customRender(
      <Modal {...defaultProps} open={false} />
    );

    const modal = queryByTestId("modal-container");
    const modalBackdrop = screen.getByTestId("modal-backdrop");
    const modalContent = screen.getByTestId("modal-content");

    expect(modalBackdrop).toHaveStyle(
      "opacity: 0; visibility: hidden; pointer-events: none;"
    );
    expect(modalContent).toHaveStyle(
      "opacity: 0; visibility: hidden; pointer-events: none;"
    );

    expect(modal).toHaveAttribute("aria-hidden", "true");
  });

  it("should call onClose when cancel button is clicked", () => {
    const { getByLabelText } = customRender(<Modal {...defaultProps} />);

    const closeButton = getByLabelText("cancel");

    fireEvent.click(closeButton);
    expect(mockOnClose).toHaveBeenCalled();
  });

  it("should call confirm when confirm button is clicked", () => {
    const { getByLabelText } = customRender(<Modal {...defaultProps} />);

    const confirmButton = getByLabelText("confirm");

    fireEvent.click(confirmButton);
    expect(mockConfirm).toHaveBeenCalled();
  });
});
