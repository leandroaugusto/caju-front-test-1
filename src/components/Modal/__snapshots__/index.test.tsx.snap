// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > should render correctly when open 1`] = `
<div>
  <div
    aria-hidden="false"
    aria-labelledby="modal-title"
    aria-modal="true"
    data-testid="modal-container"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="sc-elDIKY hBygWI"
      data-testid="modal-backdrop"
    />
    <div
      class="sc-fQpRED cLgRmg"
      data-testid="modal-content"
    >
      <p
        class="sc-dsLQwm kSsmcA"
        id="modal-title"
      >
        Mock message
      </p>
      <div
        class="sc-iKTcqh ITcZU"
      >
        <button
          aria-label="cancel"
          class="sc-gJhJTp kQOKiA"
        >
          Cancelar
        </button>
        <button
          aria-label="confirm"
          class="sc-gJhJTp hEpyQs"
        >
          Confirmar
        </button>
      </div>
    </div>
  </div>
</div>
`;
